### PR TITLE
Adding harness to full node

### PR DESCRIPTION
**Pull request by AmitShaul**
create a process to run our processes instead of k8s to do so directly

### CODE REVIEW

File: cmd/integration/harness.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+	cfg, errCfg := DefaultConfig(execPath)
+	if errCfg != nil {
+		return nil, fmt.Errorf("failed to build mock node: %v", errCfg)
+	}
+	return NewHarness(cfg, args)
+}
+
+// NewHarness creates and initializes a new instance of Harness.
+func NewHarness(cfg *ServerConfig, args []string) (*Harness, error) {
+	fmt.Println("Starting harness")
```
Comment (antonlerner) [inline cmd/integration/harness.go]: consider using spacemesh logs for consistency

File: cmd/integration/harness.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+}
+
+func main() {
+	// os.Args[0] contains the current process path
+	_, err := NewHarnessDefaultServerConfig(os.Args[1:])
+	if err != nil {
+		fmt.Println("An error has occurred while generating a new harness:", err)
+	}
+	// Sleeping here to enable the test run before harness terminates
+	fmt.Println("sleeping for 3000 seconds (50 mins)")
```
Comment (antonlerner) [inline cmd/integration/harness.go]: why terminate? or why not terminate on signal?

File: cmd/integration/harness.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+}
+
+// ProcessErrors returns a channel used for reporting any fatal process errors.
+func (h *Harness) ProcessErrors() <-chan error {
+	return h.server.errChan
+}
+
+// connectClient attempts to establish a gRPC Client connection
+// to the provided target.
+func connectClient(target string) (*grpc.ClientConn, error) {
```
Comment (antonlerner) [inline cmd/integration/harness.go]: why is this function needed? it's not used

File: cmd/integration/harness.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+		//conn:                   conn,
+		//SpacemeshServiceClient: pb.NewSpacemeshServiceClient(conn),
+	}
+
+	return h, nil
+}
+
+// TearDown stops the harness running instance.
+// The created process is killed
+func (h *Harness) TearDown() error {
```
Comment (antonlerner) [inline cmd/integration/harness.go]: Who calls this function?

File: cmd/integration/harness.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+		err = killProcess(cfg.rpcListen)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Spawn a new mockNode server process.
+	fmt.Println("harness passing the following arguments:\n", args)
+	fmt.Println("Full node server start listening on:", server.cfg.rpcListen+"\n")
+	if err := server.start(args); err != nil {
```
Comment (antonlerner) [inline cmd/integration/harness.go:83]: how does one stop the server? is there a call for it? is it needed?

File: cmd/integration/harness.go [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+		fmt.Println("Full node ERROR listening on:", server.cfg.rpcListen+"\n")
+		return nil, err
+	}
+
+	//// Verify the client connectivity.
+	//// If failed, shutdown the server.
+	//conn, err := connectClient(cfg.rpcListen)
```
Comment (antonlerner) [inline cmd/integration/harness.go]: remove unused code

File: tests/p2p/config.yaml [hunk 7]
```diff
@@ -10,8 +10,8 @@ deployment_ready_time_out: 300 #default 120
 config_path: '../config.toml'
 
 bootstrap:
-  image: 'spacemeshos/go-spacemesh:develop'
-  command: '/bin/go-p2p'
+  image: 'spacemeshos/go-spacemesh-dev:harness'
```
Comment (antonlerner) [inline tests/p2p/config.yaml]: dont push these changes. image should stay develop

File: tests/p2p/config.yaml [hunk 8]
```diff
@@ -28,11 +28,12 @@ bootstrap:
     layer-duration-sec: '60'
     layer-average-size: '20'
     hare-wakeup-delta: '10'
+    executable-path: '/bin/go-p2p'
   deployment_type: deployment # default:deployment
 # IMPORTANT: all clients on the same suite should contain the same args (hare-committee-size, hare-max-adversaries etc)!
 client:
-  image: 'spacemeshos/go-spacemesh:develop'
-  command: '/bin/go-p2p'
+  image: 'spacemeshos/go-spacemesh-dev:harness'
```
Comment (antonlerner) [inline tests/p2p/config.yaml]: dont push these changes. image should stay develop

File: tests/p2p/bootstrap_test/config.yaml [hunk 9]
```diff
@@ -10,8 +10,8 @@ deployment_ready_time_out: 120 #default 120
 config_path: '../config.toml'
 
 bootstrap:
-  image: 'spacemeshos/go-spacemesh:develop'
-  command: '/bin/go-p2p'
+  image: 'spacemeshos/go-spacemesh-dev:harness'
```
Comment (antonlerner) [inline tests/p2p/bootstrap_test/config.yaml]: dont push these changes. image should stay develop

File: tests/hare/config.yaml [hunk 10]
```diff
@@ -27,10 +27,10 @@ bootstrap:
     hare-round-duration-sec: '10'
     hare-wakeup-delta: '5'
     layer-duration-sec: '80'
-
+    executable-path: '/bin/go-hare'
 client:
-  image: 'spacemeshos/go-spacemesh:develop'
-  command: '/bin/go-hare'
+  image: 'spacemeshos/go-spacemesh-dev:harness'
```
Comment (antonlerner) [inline tests/hare/config.yaml]: dont push these changes. image should stay develop

File: tests/hare/config.yaml [hunk 11]
```diff
@@ -10,8 +10,8 @@ single_pod_ready_time_out: 60
 config_path: '../config.toml'
 
 bootstrap:
-  image: 'spacemeshos/go-spacemesh:develop'
-  command: '/bin/go-hare'
+  image: 'spacemeshos/go-spacemesh-dev:harness'
```
Comment (antonlerner) [inline tests/hare/config.yaml]: dont push these changes. image should stay develop

File: tests/config.yaml [hunk 12]
```diff
@@ -10,8 +10,8 @@ deployment_ready_time_out: 120
 config_path: '../config.toml'
 
 bootstrap:
-  image: 'spacemeshos/go-spacemesh:develop'
-  command: '/bin/go-spacemesh'
+  image: 'spacemeshos/go-spacemesh-dev:harness'
```
Comment (antonlerner) [inline tests/config.yaml]: dont push these changes. image should stay develop

File: tests/config.yaml [hunk 13]
```diff
@@ -34,9 +34,11 @@ bootstrap:
     eligibility-confidence-param: '5'
     eligibility-epoch-offset: '0'
     genesis-active-size: '50'
+    executable-path: '/bin/go-spacemesh'
+  deployment_type: deployment
 client:
-  image: 'spacemeshos/go-spacemesh:develop'
-  command: '/bin/go-spacemesh'
+  image: 'spacemeshos/go-spacemesh-dev:harness'
```
Comment (antonlerner) [inline tests/config.yaml]: dont push these changes. image should stay develop

File: cmd/integration/server.go [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+	s.cmd.Stdout = os.Stdout
+
+	// start go-spacemesh server
+	if err := s.cmd.Start(); err != nil {
+		return err
+	}
+
+	// Launch a new goroutine that bubbles up any potential fatal
+	// process errors to errChan.
+	s.processExit = make(chan struct{})
```
Comment (antonlerner) [inline cmd/integration/server.go]: who is listening to error channel?

File: cmd/integration/harness.go (hunk 2 above)
Comment (AmitShaul) [reply cmd/integration/harness.go]: fixed in cb1cd48ca644f522e9d29fefd6f714012db20427

File: cmd/integration/harness.go (hunk 3 above)
Comment (AmitShaul) [reply cmd/integration/harness.go]: for future needs, for communicating with the node.
I can remove it if you'd like it's not critical

File: cmd/integration/harness.go (hunk 4 above)
Comment (AmitShaul) [reply cmd/integration/harness.go]: we can if we'd like in the future and decide what happens when we kill the harness.
again I can remove it, not using it at the moment

File: cmd/integration/server.go (hunk 14 above)
Comment (AmitShaul) [reply cmd/integration/server.go]: fixed in 1b6eba9f2b0bfa81d747b0bc801c3f2c1fd0a2d0

File: cmd/integration/harness.go (hunk 5 above)
Comment (AmitShaul) [reply cmd/integration/harness.go:83]: maybe in the future we would like the harness to stop the sub process, at least I imagined.
right now we're not using it

File: tests/hare/config.yaml (hunk 10 above)
Comment (AmitShaul) [reply tests/hare/config.yaml]: fixed in 704b0c0b747b960ba1d5920e85d81182eaa660a8

File: tests/p2p/bootstrap_test/config.yaml (hunk 9 above)
Comment (AmitShaul) [reply tests/p2p/bootstrap_test/config.yaml]: fixed in 704b0c0b747b960ba1d5920e85d81182eaa660a8

File: tests/p2p/config.yaml (hunk 8 above)
Comment (AmitShaul) [reply tests/p2p/config.yaml]: fixed in 704b0c0b747b960ba1d5920e85d81182eaa660a8

File: tests/p2p/config.yaml (hunk 7 above)
Comment (AmitShaul) [reply tests/p2p/config.yaml]: fixed in 704b0c0b747b960ba1d5920e85d81182eaa660a8

File: cmd/integration/harness.go (hunk 6 above)
Comment (AmitShaul) [reply cmd/integration/harness.go]: fixed in 704b0c0b747b960ba1d5920e85d81182eaa660a8

File: tests/hare/config.yaml (hunk 11 above)
Comment (AmitShaul) [reply tests/hare/config.yaml]: fixed in 704b0c0b747b960ba1d5920e85d81182eaa660a8

File: tests/config.yaml (hunk 12 above)
Comment (AmitShaul) [reply tests/config.yaml]: fixed in 704b0c0b747b960ba1d5920e85d81182eaa660a8

File: tests/config.yaml (hunk 13 above)
Comment (AmitShaul) [reply tests/config.yaml]: fixed in 704b0c0b747b960ba1d5920e85d81182eaa660a8

File: cmd/integration/harness.go (hunk 4 above)
Comment (AmitShaul) [reply cmd/integration/harness.go]: fixed in 1b6eba9f2b0bfa81d747b0bc801c3f2c1fd0a2d0

File: cmd/integration/harness.go (hunk 3 above)
Comment (AmitShaul) [reply cmd/integration/harness.go]: fixed in 1b6eba9f2b0bfa81d747b0bc801c3f2c1fd0a2d0

File: cmd/integration/harness.go (hunk 1 above)
Comment (AmitShaul) [reply cmd/integration/harness.go]: fixed in 704b0c0b747b960ba1d5920e85d81182eaa660a8

File: cmd/integration/harness.go [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+// NewHarness creates and initializes a new instance of Harness.
+func NewHarness(cfg *ServerConfig, args []string) (*Harness, error) {
+	log.Info("Starting harness")
+	server, err := newServer(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// kill node process in case it is already up
+	if isListening(cfg.rpcListen) {
```
Comment (antonlerner) [inline cmd/integration/harness.go:73]: this will kill the harness. not the node process. is this expected behaviour?

File: tests/config.yaml [hunk 16]
```diff
@@ -59,6 +61,8 @@ client:
     eligibility-confidence-param: '5'
     eligibility-epoch-offset: '0'
     genesis-active-size: '50'
+    executable-path: '/bin/go-spacemesh'
```
Comment (antonlerner) [inline tests/config.yaml:64]: why is this added?

File: tests/config.yaml [hunk 17]
```diff
@@ -34,9 +34,11 @@ bootstrap:
...
```
Comment (antonlerner) [inline tests/config.yaml:37]: why is this added?

Comment (AmitShaul) [reply tests/config.yaml:37]: this goes to the harness.
it runs with various executables so every suite passes the matching one.

File: tests/config.yaml (hunk 16 above)
Comment (AmitShaul) [reply tests/config.yaml:64]: this goes to the harness.
it runs with various executables so every suite passes the matching one.

File: cmd/integration/harness.go (hunk 15 above)
Comment (AmitShaul) [reply cmd/integration/harness.go:73]: this kills any process listening on the harness port so the port will be available for the harness use.

Comment (antonlerner) [reply cmd/integration/harness.go:73]: is this code needed? think about it, you are running a single exe in a container, how can two processes exist? 
this is legacy code from the actual harness built for testing, in case we run them multiple times a harness can still be running when running a new one. this behaviour won't happen in the docker
Think whether or not you need this code here.

Comment (AmitShaul) [reply cmd/integration/harness.go:73]: removed in ac94cfaaaa15ee08f1ea691282076d8e3c81ef61

**Review by antonlerner — APPROVED** [review]

File: Makefile [hunk 18]
```diff
@@ -206,6 +214,9 @@ endif
 dockertest-sync: dockerbuild-test dockerrun-sync
 .PHONY: dockertest-sync
 
+dockertest-harness: dockerbuild-test dockerrun-harness
```
Comment (noamnelke) [inline Makefile:217]: `dockerrun-harness` doesn't exist - what is `dockertest-harness` supposed to do? Can I delete it?
